src/lib/testing/equivalent.ts

```diff
@@ -0,0 +1,242 @@
+      let x = newField(x0);
+      let y = newField(y0);
+      // outside provable code
+      handleErrors(
+        () => op1(x, y),
+        () => op2(x0, y0),
+        (a, b) => assert(a.toBigInt() === b, 'equal results')
+      );
+      handleErrors(
+        () => op1(x, y0),
+        () => op2(x0, y0),
+        (a, b) => assert(a.toBigInt() === b, 'equal results')
+      );
+      // inside provable code
+      Provable.runAndCheck(() => {
+        x = Provable.witness(Field, () => x);
+        y = Provable.witness(Field, () => y);
+        handleErrors(
+          () => op1(x, y),
+          () => op2(x0, y0),
+          (a, b) =>
+            Provable.asProver(() => assert(a.toBigInt() === b, 'equal results'))
+        );
+        handleErrors(
+          () => op1(x, y0),
+          () => op2(x0, y0),
+          (a, b) =>
+            Provable.asProver(() => assert(a.toBigInt() === b, 'equal results'))
+        );
+      });
+    });
+  }
+  function equivalentBool1(
+    op1: (x: Field) => Bool,
+    op2: (x: bigint) => boolean,
+    rng: Random<bigint> = Random.field
+  ) {
+    test(rng, (x0, assert) => {
+      let x = newField(x0);
+      // outside provable code
+      handleErrors(
+        () => op1(x),
+        () => op2(x0),
+        (a, b) => assert(a.toBoolean() === b, 'equal results')
+      );
+      // inside provable code
+      Provable.runAndCheck(() => {
+        x = Provable.witness(Field, () => x);
+        handleErrors(
+          () => op1(x),
+          () => op2(x0),
+          (a, b) =>
+            Provable.asProver(() =>
+              assert(a.toBoolean() === b, 'equal results')
+            )
+        );
+      });
+    });
+  }
+  function equivalentBool2(
+    op1: (x: Field, y: Field | bigint) => Bool,
+    op2: (x: bigint, y: bigint) => boolean,
+    rng: Random<bigint> = Random.field
+  ) {
+    test(rng, rng, (x0, y0, assert) => {
+      let x = newField(x0);
+      let y = newField(y0);
+      // outside provable code
+      handleErrors(
+        () => op1(x, y),
+        () => op2(x0, y0),
+        (a, b) => assert(a.toBoolean() === b, 'equal results')
+      );
+      handleErrors(
+        () => op1(x, y0),
+        () => op2(x0, y0),
+        (a, b) => assert(a.toBoolean() === b, 'equal results')
+      );
+      // inside provable code
+      Provable.runAndCheck(() => {
+        x = Provable.witness(Field, () => x);
+        y = Provable.witness(Field, () => y);
+        handleErrors(
+          () => op1(x, y),
+          () => op2(x0, y0),
+          (a, b) =>
+            Provable.asProver(() =>
+              assert(a.toBoolean() === b, 'equal results')
+            )
+        );
+        handleErrors(
+          () => op1(x, y0),
+          () => op2(x0, y0),
+          (a, b) =>
+            Provable.asProver(() =>
+              assert(a.toBoolean() === b, 'equal results')
+            )
+        );
+      });
+    });
+  }
+  function equivalentVoid1(
```

Comment:
nit: `unit` not `void`, see 11.3.1 of https://www.cs.cmu.edu/~rwh/pfpl/2nded.pdf . C and Java made a mistake

Reply:
fixed 🫡